src/main/java/org/javarosa/core/model/FormDef.java

```diff
@@ -1408,35 +1409,41 @@ public void readExternal(DataInputStream dis, PrototypeFactory pf) throws IOExce
      *                            (presumably in HQ) - so don't fire end of form event.
      */
     public void initialize(boolean newInstance, boolean isCompletedInstance,
-                           InstanceInitializationFactory factory) {
-        initialize(newInstance, isCompletedInstance, factory, null, false);
+                           InstanceInitializationFactory factory, CommCareSession commCareSession) {
+        initialize(newInstance, isCompletedInstance, factory, null, false, commCareSession);
     }
 
-    public void initialize(boolean newInstance, InstanceInitializationFactory factory) {
-        initialize(newInstance, false, factory, null, false);
+    public void initialize(boolean newInstance, InstanceInitializationFactory factory, CommCareSession commCareSession) {
+        initialize(newInstance, false, factory, null, false, commCareSession);
     }
 
     public void initialize(boolean newInstance, InstanceInitializationFactory factory, String locale,
-                           boolean isReadOnly) {
-        initialize(newInstance, false, factory, locale, isReadOnly);
+                           boolean isReadOnly, CommCareSession commCareSession) {
+        initialize(newInstance, false, factory, locale, isReadOnly, commCareSession);
     }
 
     /**
      * meant to be called after deserialization and initialization of handlers
      *
-     * @param newInstance true if the form is to be used for a new entry interaction,
-     *                    false if it is using an existing IDataModel
-     * @param locale      The default locale in the current environment, if provided. Can be null
-     *                    to rely on the form's internal default.
+     * @param newInstance     true if the form is to be used for a new entry interaction,
+     *                        false if it is using an existing IDataModel
+     * @param locale          The default locale in the current environment, if provided. Can be null
+     *                        to rely on the form's internal default.
+     * @param isReadOnly      If we are in read only mode and only wants to view form
+     * @param commCareSession Current commcare session the form is opened in
      */
     public void initialize(boolean newInstance, boolean isCompletedInstance,
-                           InstanceInitializationFactory factory, String locale, boolean isReadOnly) {
+                           InstanceInitializationFactory factory, String locale, boolean isReadOnly, CommCareSession commCareSession) {
         for (Enumeration en = formInstances.keys(); en.hasMoreElements(); ) {
             String instanceId = (String)en.nextElement();
             DataInstance instance = formInstances.get(instanceId);
             formInstances.put(instanceId, instance.initialize(factory, instanceId));
```

Comment:
just confirming my understanding, this will and an 'empty' instance with the same ID which then get's overridden below with the real instance from the session correct?

Reply:
This doesn't necessary have an instance with same id as the one present in the session frame xml instances. If it does, yes it will get overriden by the code below.